checkqueue: make the queue non-optional for CCheckQueueControl and drop legacy locking macro usage

theuni opened the pull request:
As part of an effort to cleanup our threading primitives and add safe `SharedMutex`/`SharedLock` impls, I'd like to get rid of the last of our legacy `ENTER_CRITICAL_SECTION`/`LEAVE_CRITICAL_SECTION` usage. This, along with a follow-up [after fixing REVERSE_LOCK](https://github.com/bitcoin/bitcoin/pull/32465) will allow us to do that.

This replaces the old macros with an RAII lock, while simplifying `CCheckQueueControl`. It now requires a `CCheckQueue`, and optionality is handled externally. In the case of validation, it is wrapped in a `std::optional`.

It also adds an `LOCK_ARGS` macro for `UniqueLock` initialization which may be helpful elsewhere.

theuni committed:
validation: only create a CCheckQueueControl if it's actually going to be used

This will allow CCheckQueueControl to require a CCheckQueue.

theuni committed:
validation: make CCheckQueueControl's CCheckQueue non-optional

This simplifies the construction logic and will allow the constructor and
destructor to lock and unlock uncondiationally.

theuni committed:
validation: use a lock for CCheckQueueControl

Uses an RAII lock for the exact same behavior as the old critical sections.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

hebasto reviewed (commented):
Concept ACK.

fjahr commented:
Concept ACK

TheCharlatan commented on src/checkqueue.h:
```diff
@@ -208,43 +208,35 @@ template <typename T, typename R = std::remove_cvref_t<decltype(std::declval<T>(
     CCheckQueueControl(const CCheckQueueControl&) = delete;
     CCheckQueueControl& operator=(const CCheckQueueControl&) = delete;
-    explicit CCheckQueueControl(CCheckQueue<T> * const pqueueIn) : pqueue(pqueueIn), fDone(false)
+    explicit CCheckQueueControl(CCheckQueue<T>& pqueueIn) : pqueue(pqueueIn), fDone(false)
```
Nit: `s/pqueue/queue/`

The relevant lines are touched already, so maybe just do it in the same commit?

fjahr replied:
double-nit: maybe `m_queue`

theuni replied:
Done.

TheCharlatan commented on src/checkqueue.h:
```diff
@@ -209,16 +209,14 @@ class CCheckQueueControl
-    {
-        ENTER_CRITICAL_SECTION(pqueue.m_control_mutex);
-    }
+    explicit CCheckQueueControl(CCheckQueue<T>& pqueueIn) : pqueue(pqueueIn), m_lock(IN_PLACE_LOCK(pqueueIn.m_control_mutex)), fDone(false) {}
```
I think it would be nice if we could get some kind of locking warning here if a second `CCheckQueueControl` is created from the same `CCheckQueue`, instead of just deadlocking. Tried coming up with a combination of LOCK and UNLOCK_FUNCTION annotation, but could not find something that actually worked.

theuni replied:
Yeah, I had hoped for the same :(

The problem here (as usual with these annotations) is aliasing.

As a simplified example:
```c++
CCheckQueueControl<CScriptCheck> control(m_chainman.GetCheckQueue());
CCheckQueueControl<CScriptCheck> control2(m_chainman.GetCheckQueue());
```

Here, clang has no idea that `&control.m_lock.m_control_mutex` == `&control2.m_lock.m_control_mutex`.

The only alternative, really, would be something like:

```diff
diff --git a/src/checkqueue.h b/src/checkqueue.h
index ade793ab392..5249e040d24 100644
--- a/src/checkqueue.h
+++ b/src/checkqueue.h
@@ -188,6 +188,11 @@ public:
         }
     }

+    UniqueLock<Mutex> TryLock()
+    {
+        return {IN_PLACE_TRY_LOCK(m_control_mutex)};
+    }
+
     ~CCheckQueue()
     {
         WITH_LOCK(m_mutex, m_request_stop = true);
@@ -209,13 +214,14 @@ class CCheckQueueControl
 {
 private:
     CCheckQueue<T, R>& pqueue;
+    UniqueLock<Mutex> m_lock;
     bool fDone;

 public:
     CCheckQueueControl() = delete;
     CCheckQueueControl(const CCheckQueueControl&) = delete;
     CCheckQueueControl& operator=(const CCheckQueueControl&) = delete;
-    explicit CCheckQueueControl(CCheckQueue<T>& pqueueIn) : pqueue(pqueueIn), fDone(false)
+    explicit CCheckQueueControl(CCheckQueue<T>& pqueueIn, UniqueLock<Mutex>&& lock) : pqueue(pqueueIn), m_lock(std::move(lock)), fDone(false)
     {
         ENTER_CRITICAL_SECTION(pqueue.m_control_mutex);
     }
diff --git a/src/sync.h b/src/sync.h
index b22956ef1ab..db4fd977cee 100644
--- a/src/sync.h
+++ b/src/sync.h
@@ -193,6 +193,8 @@ public:
             Enter(pszName, pszFile, nLine);
     }

+    UniqueLock(UniqueLock&&) = default;
+
     ~UniqueLock() UNLOCK_FUNCTION()
     {
         if (Base::owns_lock())
@@ -260,6 +262,7 @@ inline MutexType* MaybeCheckNotHeld(MutexType* m) LOCKS_EXCLUDED(m) LOCK_RETURNE
     UniqueLock criticalblock2(MaybeCheckNotHeld(cs2), #cs2, __FILE__, __LINE__)
 #define TRY_LOCK(cs, name) UniqueLock name(MaybeCheckNotHeld(cs), #cs, __FILE__, __LINE__, true)
 #define WAIT_LOCK(cs, name) UniqueLock name(MaybeCheckNotHeld(cs), #cs, __FILE__, __LINE__)
+#define IN_PLACE_TRY_LOCK(cs) MaybeCheckNotHeld(cs), #cs, __FILE__, __LINE__, true

 #define ENTER_CRITICAL_SECTION(cs)                            \
     {                                                         \
diff --git a/src/validation.cpp b/src/validation.cpp
index 3fa7afe3090..d690191b38e 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2612,7 +2612,10 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
     // doesn't invalidate pointers into the vector, and keep txsdata in scope
     // for as long as `control`.
     std::optional<CCheckQueueControl<CScriptCheck>> control;
-    if (fScriptChecks && parallel_script_checks) control.emplace(m_chainman.GetCheckQueue());
+    auto& queue = m_chainman.GetCheckQueue();
+    auto queue_lock = queue.TryLock();
+    assert(queue_lock);
+    if (fScriptChecks && parallel_script_checks) control.emplace(m_chainman.GetCheckQueue(), std::move(queue_lock));

     std::vector<PrecomputedTransactionData> txsdata(block.vtx.size());

```

Unfortunately, `try_lock` is allowed to fail spuriously, so that's not safe :(

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/32467#discussion_r2083592266

Following seems to prevent duplicate locking. Error message is:

```c++
src/bench/checkqueue.cpp:60:42: error: acquiring mutex 'queue.m_control_mutex' that is already held [-Werror,-Wthread-safety-analysis]
   60 |         CCheckQueueControl<PrevectorJob> control2(queue);
      |                                          ^
src/bench/checkqueue.cpp:59:42: note: mutex acquired here
   59 |         CCheckQueueControl<PrevectorJob> control(queue);
      |                                          ^
1 error generated.
```
<details><summary>diff</summary>
<p>

```diff
--- a/src/bench/checkqueue.cpp
+++ b/src/bench/checkqueue.cpp
@@ -57,6 +57,7 @@ static void CCheckQueueSpeedPrevectorJob(benchmark::Bench& bench)
     bench.minEpochIterations(10).batch(BATCH_SIZE * BATCHES).unit("job").run([&] {
         // Make insecure_rand here so that each iteration is identical.
         CCheckQueueControl<PrevectorJob> control(queue);
+        CCheckQueueControl<PrevectorJob> control2(queue);
         for (auto vChecks : vBatches) {
             control.Add(std::move(vChecks));
         }
--- a/src/checkqueue.h
+++ b/src/checkqueue.h
@@ -205,7 +205,7 @@ public:
  * queue is finished before continuing.
  */
 template <typename T, typename R = std::remove_cvref_t<decltype(std::declval<T>()().value())>>
-class CCheckQueueControl
+class SCOPED_LOCKABLE CCheckQueueControl
 {
 private:
     CCheckQueue<T, R>& m_queue;
@@ -216,7 +216,7 @@ public:
     CCheckQueueControl() = delete;
     CCheckQueueControl(const CCheckQueueControl&) = delete;
     CCheckQueueControl& operator=(const CCheckQueueControl&) = delete;
-    explicit CCheckQueueControl(CCheckQueue<T>& queueIn) : m_queue(queueIn), m_lock(IN_PLACE_LOCK(queueIn.m_control_mutex)), fDone(false) {}
+    explicit CCheckQueueControl(CCheckQueue<T>& queueIn) EXCLUSIVE_LOCK_FUNCTION(queueIn.m_control_mutex) : m_queue(queueIn), m_lock(IN_PLACE_LOCK(queueIn.m_control_mutex)), fDone(false) {}
 
     std::optional<R> Complete()
     {
@@ -230,7 +230,7 @@ public:
         m_queue.Add(std::move(vChecks));
     }
 
-    ~CCheckQueueControl()
+    ~CCheckQueueControl() UNLOCK_FUNCTION()
     {
         if (!fDone)
             Complete();
```
</p>
</details>

theuni replied:
Nice! I guess it manages to see through the references after all. My testing shows that:
```c++
    auto& queue = m_chainman.GetCheckQueue();
    CCheckQueueControl<CScriptCheck> control(queue);
    CCheckQueueControl<CScriptCheck> control2(queue)
```

```c++
    CCheckQueueControl<CScriptCheck> control(m_chainman.GetCheckQueue());
    CCheckQueueControl<CScriptCheck> control2(m_chainman.GetCheckQueue());
```

Both of those indeed work.

I suppose I came to the conclusion that the `SCOPED_LOCKABLE` approach didn't help because neither of the above work when using a `std::optional`. So it won't actually work in the validation case.

But still better than nothing. Will add. Thanks!

sedited reviewed (approved):
ACK 5aca850c205a20a0f198827f9797fb8053f2b3dd

fjahr commented:
Code review ACK 5aca850c205a20a0f198827f9797fb8053f2b3dd

theuni commented:
Updated to address @TheCharlatan's nit.

sedited reviewed (approved):
ACK da5d55227e6aead8a9fbb85e967801cfcd4283ce

fjahr commented:
re-ACK da5d55227e6aead8a9fbb85e967801cfcd4283ce

ryanofsky commented on src/validation.cpp:
```diff
@@ -2680,7 +2682,8 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
                 break;
             }
-            control.Add(std::move(vChecks));
+            if (control)
+                control->Add(std::move(vChecks));
```
In commit "validation: only create a CCheckQueueControl if it's actually going to be used" (bcde9153012b2ce3e9ff5b7cd1c6f9866453e727)

Developer notes want single-line if or multiple line if with braces, not multiline if without braces.

>  - If an `if` only has a single-statement `then`-clause, it can appear
>    on the same line as the `if`, without braces. In every other case,
>    braces are required, and the `then` and `else` clauses must appear
>    correctly indented on a new line.

ryanofsky commented on src/validation.cpp:
```diff
@@ -2680,7 +2682,8 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
                 break;
             }
-            control.Add(std::move(vChecks));
+            if (control)
```
In commit "validation: only create a CCheckQueueControl if it's actually going to be used" (bcde9153012b2ce3e9ff5b7cd1c6f9866453e727)

Seems like it would be good to change `parallel_script_checks ? &vChecks : nullptr` on line 2679 above to `control ? &vChecks : nullptr` to be checking for the `control` condition consistently in this function.

IMO, it would also be nice to have a comment here saying that if `control` is unset, checks were already performed in the `CheckInputScripts` call. Otherwise it looks like they are being skipped here.

theuni replied:
Mmm, this is kinda awkward either way. It'd be better imo to pass the actual `CCheckQueueControl&` as an argument to `CheckInputScripts`, but that makes this much more complicated for a bunch of reasons.

I pushed an additional commit to try to make this a little easier to read. Please let me know what you think. If it's over-complicating what's otherwise a pretty trivial PR, I can remove it.

ryanofsky commented on src/sync.h:
```diff
@@ -260,6 +260,7 @@ inline MutexType* MaybeCheckNotHeld(MutexType* m) LOCKS_EXCLUDED(m) LOCK_RETURNE
     UniqueLock criticalblock2(MaybeCheckNotHeld(cs2), #cs2, __FILE__, __LINE__)
 #define TRY_LOCK(cs, name) UniqueLock name(MaybeCheckNotHeld(cs), #cs, __FILE__, __LINE__, true)
 #define WAIT_LOCK(cs, name) UniqueLock name(MaybeCheckNotHeld(cs), #cs, __FILE__, __LINE__)
+#define IN_PLACE_LOCK(cs) MaybeCheckNotHeld(cs), #cs, __FILE__, __LINE__
```
In commit "validation: use a lock for CCheckQueueControl" (da5d55227e6aead8a9fbb85e967801cfcd4283ce)

Something like LOCK_ARGS might be a better name for this macro than than IN_PLACE_LOCK, because it expands to a list of 4 arguments, not a lock. Also, TRY_LOCK and WAIT_LOCK above could be deduplicated and written using LOCK_ARGS.

I don't think there is a close parallel between IN_PLACE_LOCK and std::in_place other than that they both get used as constructor arguments. IN_PLACE_LOCK isn't being used to help construct objects in place inside a larger std::any/std::optional/std::variant object like std::in_place is

theuni replied:
Done. I wasn't sure if `__LINE__` would work correctly when called from another macro, but it seems to do the right thing :)

ryanofsky reviewed (approved):
Code review ACK da5d55227e6aead8a9fbb85e967801cfcd4283ce. This looks good and is a nice improvement. Left some suggestions below I think it could be good to follow up on, though

theuni committed:
threading: add LOCK_ARGS macro

This is useful for initializing locks in a constructor.

theuni committed:
validation: clean up and clarify CheckInputScripts logic

CheckInputScripts behaves differently depending on whether or not it was called
with a vector for checks. Make this difference clear by calling it differently
depending on whether or not control exists. Though more verbose, it should be
more straightforward to understand what's happening this way.

Also remove parallel_script_checks, as `if(control)` is a better test.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

theuni commented:
Updated to resolve @ryanofsky's comments. I added a new comment to (hopefully) clarify the `CheckInputScripts` calling.

hebasto reviewed (commented):
Approach ACK 833dc7b9e3672b9a9033d02802ac60a48d4a8b3c, I have reviewed the code and it looks OK. Still testing.

> It also adds an `IN_PLACE_LOCK` macro for `UniqueLock` initialization which may be helpful elsewhere.

Update the macro name in the PR description?

ryanofsky commented on src/validation.cpp:
```diff
@@ -2671,18 +2670,22 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
+            // they need to be added to control which runs them asynchronously. Otherwise, CheckInputScripts runs the checks before returning.
+            if (control) {
+                std::vector<CScriptCheck> vChecks;
+                Assert(CheckInputScripts(tx, tx_state, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache, &vChecks));
```
In commit "validation: clean up and clarify CheckInputScripts logic" (833dc7b9e3672b9a9033d02802ac60a48d4a8b3c)

Not important, but I think it is not ideal for CheckInputScripts return value to be handled inconsistently and sometimes abort. It makes it more difficult for CheckInputScripts to change in the future (to add checks or shift between delayed and upfront checking) if one caller is making stronger assumptions than the other callers about how it works and what it may return. Would suggest treating both cases here more similarly:

```diff
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2673,14 +2673,18 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
         if (!tx.IsCoinBase() && fScriptChecks)
         {
             bool fCacheResults = fJustCheck; /* Don't cache results if we're actually connecting blocks (still consult the cache, though) */
+            bool tx_ok;
             TxValidationState tx_state;
             // If CheckInputScripts is called with a pointer to a checks vector, the resulting checks are appended to it. In that case
             // they need to be added to control which runs them asynchronously. Otherwise, CheckInputScripts runs the checks before returning.
             if (control) {
                 std::vector<CScriptCheck> vChecks;
-                Assert(CheckInputScripts(tx, tx_state, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache, &vChecks));
+                tx_ok = CheckInputScripts(tx, tx_state, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache, &vChecks);
                 control->Add(std::move(vChecks));
-            } else if (!CheckInputScripts(tx, tx_state, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache)) {
+            } else {
+                tx_ok = CheckInputScripts(tx, tx_state, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache);
+            }
+            if (!tx_ok) {
                 // Any transaction validation failure in ConnectBlock is a block consensus failure
                 state.Invalid(BlockValidationResult::BLOCK_CONSENSUS,
                               tx_state.GetRejectReason(), tx_state.GetDebugMessage());
```

theuni replied:
Yeah, I had the same thought but I guess I came to a different conclusion. I figured if a false return only has meaning in the sync code-path, the true return should be enforced in the async as well. That way if someone ever tries to add a `return false` in the batched case, we'd catch it right away in testing.

ryanofsky replied:
> That way if someone ever tries to add a `return false` in the batched case, we'd catch it right away in testing.

Catch what though? Catch CheckInputScripts detecting an invalid input and returning false like someone would reasonably expect it to? There aren't warnings or any indication inside CheckInputScripts saying that returning false may crash the program. So I don't think it's hard to imagine someone optimizing CheckInputScripts or refactoring or just adding a new check that returns false. They could even write test code that triggers the check but bypasses this particular call path. And the check could avoid initially causing errors in production because it would normally pass. But at some point later the check could be triggered and cause crashes, and the diff I suggested above might be the logical fix.

Perhaps this is not realistic, but naively to me it would seem safest to handle CheckInputScripts failing in ConnectBlock by treating the block as invalid (as the code is currently doing) rather than crashing.

If it is also important to ensure that CheckInputScripts returns true when CheckQueue::HasThreads is true, it seems like it would be safer to check for that with an Assume+LogWarning rather than an Assert.

I'm not sure about this and maybe I am thinking about it in the wrong way, but that's my logic here.

fjahr replied:
This whole conversation reminded me that I really wanted to split up `CheckInputScripts` after interacting with it for the batch validation stuff. I will open a draft of that for conceptual feedback.

ryanofsky reviewed (approved):
Code review ACK 833dc7b9e3672b9a9033d02802ac60a48d4a8b3c. Main changes since last review were adding lock annotations and a new commit to clarify CheckInputScripts.

I definitely think the new commit 833dc7b9e3672b9a9033d02802ac60a48d4a8b3c is good and makes code less confusing. I left one suggestion about it, but it is not important.

fjahr commented:
re-ACK fd290730f530a8b76a9607392f49830697cdd7c5

ryanofsky reviewed (approved):
Code review ACK fd290730f530a8b76a9607392f49830697cdd7c5, just removing assert since last review. Thanks for considering all the comments and feedback!

sedited reviewed (approved):
Re-ACK fd290730f530a8b76a9607392f49830697cdd7c5

mzumsande commented on src/validation.cpp:
```diff
@@ -2671,18 +2670,26 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
+            if (control) {
+                std::vector<CScriptCheck> vChecks;
+                tx_ok = CheckInputScripts(tx, tx_state, view, flags, fCacheResults, fCacheResults, txsdata[i], m_chainman.m_validation_cache, &vChecks);
+                if (tx_ok) control->Add(std::move(vChecks));
```
just noting that `CheckInputScripts()` will always return `true` in the multi-threaded case (it doesn't do any validation after all), so this logic could be simpler / `tx_ok` isn't really necessary.

theuni replied:
@mzumsande See the (hidden) discussion here: https://github.com/bitcoin/bitcoin/pull/32467#discussion_r2098305199

This was mostly added to prevent future footguns if an early false return is ever added. But yeah, it's meaningless and confusing as-is, and prompted @fjahr's PR above :)

mzumsande reviewed (commented):
post-merge code review ACK fd290730f530a8b76a9607392f49830697cdd7c5

davidgumberg commented on src/validation.cpp:
```diff
@@ -2612,7 +2611,7 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
     // for as long as `control`.
     std::optional<CCheckQueueControl<CScriptCheck>> control;
-    if (fScriptChecks && parallel_script_checks) control.emplace(m_chainman.GetCheckQueue());
+    if (auto& queue = m_chainman.GetCheckQueue(); queue.HasThreads() && fScriptChecks) control.emplace(queue);
```
Dumb question: Is there a reason why the order was changed here between `HasThreads()` and `fSCriptChecks`? Is it because `HasThreads()` is cheap and the line is more readible with the check next to the initializer?

theuni replied:
Yep, no reason other than I thought it was more readable that way.

davidgumberg commented:
post-merge crACK https://github.com/bitcoin/bitcoin/commit/fd290730f530a8b76a9607392f49830697cdd7c5
